File: frontend/src/api/inventory.ts
import { apiClient } from './client';

export interface PaginatedResponse<T> {
  items: T[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

export interface InventoryItemResponse {
  id: number;
  name: string;
  sku: string;
  barcode?: string;
  category_id?: number;
  item_type: string;
  current_quantity: number;
  reserved_quantity: number;
  minimum_quantity: number;
  bin_id?: number;
  description?: string;
  image_url?: string;
  parent_id?: number;
  children?: InventoryItemResponse[];
  is_active: boolean;
  created_at: string;
  updated_at: string;
  deleted_at?: string;
}

export interface InventoryItemChildRequest {
  name: string;
  sku: string;
  barcode?: string;
  item_type?: string;
  current_quantity?: number;
  minimum_quantity?: number;
  description?: string;
  primary_attribute?: string;
  secondary_attribute?: string;
  notes?: string;
}

export interface InventoryItemBatchCreateRequest {
  parent: InventoryItemCreateRequest;
  children: InventoryItemChildRequest[];
}

export interface InventoryItemCreateRequest {
  name: string;
  sku: string;
  barcode?: string;
  category_id?: number;
  item_type: string;
  current_quantity: number;
  minimum_quantity: number;
  bin_id?: number;
  description?: string;
  image_url?: string;
  parent_id?: number;
}

export interface InventoryItemUpdateRequest {
  name?: string;
  category_id?: number;
  item_type?: string;
  minimum_quantity?: number;
  bin_id?: number;
  description?: string;
  image_url?: string;
  parent_id?: number;
  current_quantity?: number;
  barcode?: string;
}

export interface InventoryTransactionResponse {
  id: number;
  transaction_type: string;
  previous_quantity: number;
  change_quantity: number;
  new_quantity: number;
  reference_type?: string;
  reference_id?: number;
  reason?: string;
  created_by?: number;
  created_at: string;
}

export interface InventoryItemDetailResponse extends InventoryItemResponse {
  transactions: InventoryTransactionResponse[];
  serial_numbers?: SerialNumberResponse[];
  parent?: InventoryItemResponse;
}

export interface SerialNumberResponse {
  id: number;
  item_id: number;
  serial_number: string;
  batch_id?: string;
  unit_condition: string;
  location_bin_id?: number;
  assigned_to_order_id?: number;
  notes?: string;
  created_at: string;
  updated_at: string;
}

export interface RestockRequest {
  item_id: number;
  quantity: number;
  reason: string;
}

export interface AdjustmentRequest {
  item_id: number;
  new_quantity: number;
  reason: string;
}

export const inventoryApi = {
  list: async (
    page: number = 1,
    size: number = 20,
    search?: string,
    category_id?: number,
    item_type?: string,
    low_stock: boolean = false
  ): Promise<PaginatedResponse<InventoryItemResponse>> => {
    const params = new URLSearchParams();
    params.append('page', page.toString());
    params.append('size', size.toString());
    if (search) {
      params.append('search', search);
    }
    if (category_id) {
      params.append('category_id', category_id.toString());
    }
    if (item_type) {
      params.append('item_type', item_type);
    }
    if (low_stock) {
      params.append('low_stock', 'true');
    }
    const response = await apiClient.get<PaginatedResponse<InventoryItemResponse>>(
      `/inventory/items?${params.toString()}`
    );
    return response.data;
  },

  get: async (itemId: number): Promise<InventoryItemDetailResponse> => {
    const response = await apiClient.get<InventoryItemDetailResponse>(`/inventory/items/${itemId}`);
    return response.data;
  },

  create: async (
    data: InventoryItemCreateRequest
  ): Promise<InventoryItemResponse> => {
    const response = await apiClient.post<InventoryItemResponse>('/inventory/items', data);
    return response.data;
  },

  createBatch: async (
    data: InventoryItemBatchCreateRequest
  ): Promise<InventoryItemResponse> => {
    const response = await apiClient.post<InventoryItemResponse>('/inventory/items/batch', data);
    return response.data;
  },

  update: async (
    itemId: number,
    data: InventoryItemUpdateRequest
  ): Promise<InventoryItemResponse> => {
    const response = await apiClient.patch<InventoryItemResponse>(
      `/inventory/items/${itemId}`,
      data
    );
    return response.data;
  },

  delete: async (itemId: number): Promise<void> => {
    await apiClient.delete(`/inventory/items/${itemId}`);
  },

  getByBarcode: async (barcode: string): Promise<InventoryItemResponse> => {
    const response = await apiClient.get<InventoryItemResponse>(
      `/inventory/items/barcode/${barcode}`
    );
    return response.data;
  },

  getTransactions: async (itemId: number): Promise<InventoryTransactionResponse[]> => {
    const response = await apiClient.get<InventoryTransactionResponse[]>(
      `/inventory/items/${itemId}/transactions`
    );
    return response.data;
  },

  restock: async (data: RestockRequest): Promise<InventoryTransactionResponse> => {
    const response = await apiClient.post<InventoryTransactionResponse>(
      '/inventory/restock',
      data
    );
    return response.data;
  },

  adjust: async (data: AdjustmentRequest): Promise<InventoryTransactionResponse> => {
    const response = await apiClient.post<InventoryTransactionResponse>(
      '/inventory/adjust',
      data
    );
    return response.data;
  },

  getSerials: async (itemId: number): Promise<SerialNumberResponse[]> => {
    const response = await apiClient.get<SerialNumberResponse[]>(`/inventory/${itemId}/serials`);
    return response.data;
  },

  createSingleSerials: async (itemId: number, payload: { count: number; base_serial?: string; batch_id?: string; condition?: string }): Promise<SerialNumberResponse[]> => {
    const response = await apiClient.post<SerialNumberResponse[]>(`/inventory/${itemId}/serials/single`, payload);
    return response.data;
  },

  createRangeSerials: async (itemId: number, payload: { start_serial: string; end_serial: string; batch_id?: string; condition?: string }): Promise<SerialNumberResponse[]> => {
    const response = await apiClient.post<SerialNumberResponse[]>(`/inventory/${itemId}/serials/range`, payload);
    return response.data;
  },

  importSerials: async (itemId: number, payload: { serials: string[]; batch_id?: string; condition?: string }): Promise<SerialNumberResponse[]> => {
    const response = await apiClient.post<SerialNumberResponse[]>(`/inventory/${itemId}/serials/import`, payload);
    return response.data;
  },

  deleteSerial: async (itemId: number, serialId: number): Promise<void> => {
    await apiClient.delete(`/inventory/${itemId}/serials/${serialId}`);
  },

  updateSerial: async (itemId: number, serialId: number, data: { unit_condition?: string; notes?: string }): Promise<SerialNumberResponse> => {
    const response = await apiClient.patch<SerialNumberResponse>(`/inventory/${itemId}/serials/${serialId}`, data);
    return response.data;
  },
};

File: backend/app/schemas/inventory.py
from __future__ import annotations
from typing import Optional, List
from pydantic import BaseModel, ConfigDict, computed_field, field_validator
from datetime import datetime


# ============ INVENTORY CATEGORIES ============
class InventoryCategoryBase(BaseModel):
    name: str
    parent_id: Optional[int] = None


class InventoryCategoryCreate(InventoryCategoryBase):
    pass


class InventoryCategoryResponse(InventoryCategoryBase):
    id: int

    model_config = ConfigDict(from_attributes=True)


# ============ INVENTORY ITEMS ============
class InventoryItemBase(BaseModel):
    name: str
    sku: str
    erp_number: Optional[str] = None
    barcode: Optional[str] = None
    category_id: Optional[int] = None
    item_type: str = "consumable"
    minimum_quantity: float = 0
    bin_id: Optional[int] = None
    description: Optional[str] = None
    image_url: Optional[str] = None
    parent_id: Optional[int] = None
    is_container: bool = False
    attributes: Optional[dict] = None


class InventoryItemCreate(InventoryItemBase):
    current_quantity: float = 0
    is_draft: bool = False  # Save as draft instead of publishing immediately


class InventoryItemChildCreate(BaseModel):
    name: str
    sku: str
    barcode: Optional[str] = None
    item_type: str = "consumable"
    current_quantity: float = 0
    minimum_quantity: float = 0
    description: Optional[str] = None
    primary_attribute: Optional[str] = None
    secondary_attribute: Optional[str] = None
    notes: Optional[str] = None


class InventoryItemBatchCreate(BaseModel):
    parent: InventoryItemCreate
    children: List[InventoryItemChildCreate] = []


class InventoryItemUpdate(BaseModel):
    name: Optional[str] = None
    category_id: Optional[int] = None
    item_type: Optional[str] = None
    minimum_quantity: Optional[float] = None
    bin_id: Optional[int] = None
    description: Optional[str] = None
    image_url: Optional[str] = None
    parent_id: Optional[int] = None
    is_container: Optional[bool] = None
    attributes: Optional[dict] = None
    expiry_date: Optional[datetime] = None
    allow_no_expiry: Optional[bool] = None
    stock_status: Optional[str] = None  # active | expired | damaged
    current_quantity: Optional[float] = None  # Allow direct quantity edits


class InventoryTransactionResponse(BaseModel):
    id: int
    transaction_type: str
    previous_quantity: float
    change_quantity: float
    new_quantity: float
    reference_type: Optional[str] = None
    reference_id: Optional[int] = None
    reason: Optional[str] = None
    created_by: Optional[int] = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class InventoryItemResponse(InventoryItemBase):
    id: int
    current_quantity: float
    reserved_quantity: float
    is_active: bool
    is_container: bool = False
    is_draft: bool = False
    expiry_date: Optional[datetime] = None
    allow_no_expiry: bool = True
    stock_status: str = "active"  # active | expired | damaged
    created_at: datetime
    updated_at: datetime
    children: List['InventoryItemResponse'] = []
    attributes: Optional[dict] = None
    front_image_url: Optional[str] = None
    back_image_url: Optional[str] = None

    model_config = ConfigDict(from_attributes=True, arbitrary_types_allowed=True)

    @field_validator('attributes', mode='before')
    @classmethod
    def convert_attributes(cls, v):
        """Convert ORM attributes list to dict."""
        if isinstance(v, list) and v:
            # Convert list of ORM objects to dict
            return {attr.attribute_name: attr.attribute_value for attr in v}
        return v or None

    @computed_field
    @property
    def available_quantity(self) -> float:
        if self.is_container and self.children:
            return sum(c.available_quantity for c in self.children)
        return self.current_quantity - self.reserved_quantity


class InventoryItemDetailResponse(InventoryItemResponse):
    transactions: List[InventoryTransactionResponse] = []
    images: List['InventoryItemImageResponse'] = []
    serial_numbers: List['SerialNumberResponse'] = []
    parent: Optional['InventoryItemResponse'] = None


# ============ INVENTORY ITEM IMAGES ============
class InventoryItemImageCreate(BaseModel):
    image_type: str


class InventoryItemImageResponse(BaseModel):
    id: int
    item_id: int
    image_type: str
    image_url: str
    uploaded_by: Optional[int] = None
    uploaded_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ============ SERIAL NUMBERS ============
class SerialNumberCreate(BaseModel):
    count: int = 1
    base_serial: Optional[str] = None
    batch_id: Optional[str] = None
    condition: str = "new"


class SerialNumberBatchCreate(BaseModel):
    start_serial: str
    end_serial: str
    batch_id: Optional[str] = None
    condition: str = "new"


class SerialNumberImportCreate(BaseModel):
    serials: List[str]
    batch_id: Optional[str] = None
    condition: str = "new"


class SerialNumberUpdate(BaseModel):
    unit_condition: Optional[str] = None
    location_bin_id: Optional[int] = None
    assigned_to_order_id: Optional[int] = None
    notes: Optional[str] = None


class SerialNumberResponse(BaseModel):
    id: int
    item_id: int
    serial_number: str
    batch_id: Optional[str] = None
    unit_condition: str = "new"
    location_bin_id: Optional[int] = None
    assigned_to_order_id: Optional[int] = None
    notes: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ============ RESTOCK & ADJUST ============
class RestockRequest(BaseModel):
    item_id: int
    quantity: float
    reason: str


class AdjustmentRequest(BaseModel):
    item_id: int
    new_quantity: float
    reason: str


# Rebuild models that use forward references to ensure they can find the classes defined below them
# This is necessary for Pydantic v2 + FastAPI route registration
InventoryItemDetailResponse.model_rebuild()
InventoryItemResponse.model_rebuild()
